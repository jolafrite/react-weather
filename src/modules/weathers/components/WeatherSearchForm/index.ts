export * from './WeatherSearchForm';
export { default } from './WeatherSearchForm';
