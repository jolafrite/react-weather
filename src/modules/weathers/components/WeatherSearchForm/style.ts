import { makeStyles, createStyles } from "@material-ui/core";

export const useStyles = makeStyles(() =>
  createStyles({
    root: {
      display: 'flex',
      alignItems: 'center',
      padding: '2px 4px',
      width: 400,
    },
    input: {
      marginLeft: 8,
      flex: 1,
    },
    iconButton: {
      padding: 10,
    },
    progress: {

    }
  })
);
