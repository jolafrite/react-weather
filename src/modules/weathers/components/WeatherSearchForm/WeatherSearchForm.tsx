import React, { useState, ChangeEvent, FormEvent, memo } from "react";
import { useStyles } from "./style";
import { IconButton, InputBase, CircularProgress } from "@material-ui/core";
import { Search } from '@material-ui/icons';

export interface IWeatherSearchFormProps {
  loading: boolean;
  onSubmit: (query: string) => void;
}

const WeatherSearchForm: React.FC<IWeatherSearchFormProps> = props => {
  const classes = useStyles();
  const { loading, onSubmit } = props;
  const [query, setQuery] = useState('');

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => setQuery(e.target.value);
  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    onSubmit(query);
  }

  const disableButton = query.trim() === '' || loading;

  return (
    <form
      onSubmit={handleSubmit}
      className={classes.root}
    >
      <InputBase
        className={classes.input}
        placeholder="Enter the city name"
        inputProps={{ 'aria-label': 'Enter the city name' }}
        required
        value={query}
        onChange={handleChange}
      />

      <IconButton
        type="submit"
        className={classes.iconButton}
        aria-label="Search"
        disabled={disableButton}
      >
        {!loading
          ? <Search />
          : <CircularProgress
              size={24}
              className={classes.progress}
            />
        }
      </IconButton>
    </form>
  )
};

export default memo(WeatherSearchForm);
