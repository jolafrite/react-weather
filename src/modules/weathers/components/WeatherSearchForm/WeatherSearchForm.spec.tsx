import React from 'react';
import WeatherSearchForm, { IWeatherSearchFormProps } from "./WeatherSearchForm";
import {render, cleanup, fireEvent} from '@testing-library/react';

describe('component: WeatherSearchForm', () => {

  const createComponent = (values = {}) => {
    const props: IWeatherSearchFormProps = {
      loading: false,
      onSubmit: () => {},
      ...values,
    };

    return render(
      <WeatherSearchForm
        {...props}
      />
    );
  };

  afterEach(cleanup);

  describe('init', () => {
    test('should only render the form', () => {
      const component = createComponent();

      expect(
        component.container
      ).toMatchSnapshot();
    });
  });

  describe('set a value into the input', () => {
    test('should enable the button', () => {
      const component = createComponent();

      const input = component.getByPlaceholderText("Enter the city name");
      fireEvent.change(input, { target: { value: 'Paris' } })

      expect(
        component.container
      ).toMatchSnapshot();
    });
  });

  describe('loading is set to true', () => {
    test('should disable the button and replace the icon with a spinner', () => {
      const component = createComponent({ loading: true });

      expect(
        component.container
      ).toMatchSnapshot();
    })
  });

});
