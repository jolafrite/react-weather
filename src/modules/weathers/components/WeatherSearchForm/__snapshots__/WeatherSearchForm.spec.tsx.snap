// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: WeatherSearchForm init should only render the form 1`] = `
<div>
  <form
    class="makeStyles-root-1"
  >
    <div
      class="MuiInputBase-root makeStyles-input-2"
      required=""
    >
      <input
        aria-label="Enter the city name"
        class="MuiInputBase-input"
        placeholder="Enter the city name"
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      aria-label="Search"
      class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-3 Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="submit"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </span>
    </button>
  </form>
</div>
`;

exports[`component: WeatherSearchForm loading is set to true should disable the button and replace the icon with a spinner 1`] = `
<div>
  <form
    class="makeStyles-root-93"
  >
    <div
      class="MuiInputBase-root makeStyles-input-94"
      required=""
    >
      <input
        aria-label="Enter the city name"
        class="MuiInputBase-input"
        placeholder="Enter the city name"
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      aria-label="Search"
      class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-95 Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="submit"
    >
      <span
        class="MuiIconButton-label"
      >
        <div
          class="MuiCircularProgress-root makeStyles-progress-96 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
          role="progressbar"
          style="width: 24px; height: 24px;"
        >
          <svg
            class="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
      </span>
    </button>
  </form>
</div>
`;

exports[`component: WeatherSearchForm set a value into the input should enable the button 1`] = `
<div>
  <form
    class="makeStyles-root-42"
  >
    <div
      class="MuiInputBase-root makeStyles-input-43"
      required=""
    >
      <input
        aria-label="Enter the city name"
        class="MuiInputBase-input"
        placeholder="Enter the city name"
        required=""
        type="text"
        value="Paris"
      />
    </div>
    <button
      aria-label="Search"
      class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-44"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </form>
</div>
`;
